File: src/routes/payments.js
// ============================================================================
// PAYMENTS ROUTE — M-PESA Integration & Payment Management
// ============================================================================
// This file handles EVERYTHING related to money:
//   1. M-PESA Validation URL  — Safaricom asks "is this account valid?" BEFORE taking money
//   2. M-PESA Confirmation URL — Safaricom says "payment done" → we record it automatically
//   3. Manual payment entry    — For cash, bank transfer, etc.
//   4. Payment history lookup  — See all payments for a house
// ============================================================================

// --- IMPORTS ---
// express = the web framework. Router = mini-app for handling routes under /api/payments
const express = require('express');
const router = express.Router();

// v4:uuidv4 = generates unique IDs like "550e8400-e29b-41d4-a716-446655440000"
// We use these for every new record (payment, bill, etc.)
const { v4: uuidv4 } = require('uuid');

// db = our PostgreSQL connection pool. 
// db.query() sends SQL to the database and returns results
const db = require('../config/database');

// company = holds company details (name, paybill number, phone) 
// and SMS template functions (getBillSms, getPaymentConfirmationSms, etc.)
const company = require('../config/company');

// authenticate = middleware that checks JWT token on protected routes
// authorize = checks if user has the right role (super_admin, admin, clerk)
const { authenticate, authorize } = require('../middleware/auth');

// ============================================================================
// ROUTE 1: M-PESA VALIDATION URL (POST /api/payments/validate)
// ============================================================================
// WHEN DOES THIS FIRE?
//   When a resident opens M-PESA → Lipa na M-PESA → Paybill → enters account number
//   Safaricom sends a request HERE before taking the money
// 
// WHAT DOES IT DO?
//   1. Receives the account number (BillRefNumber) from Safaricom
//   2. Checks if that account exists in our database
//   3. Checks if the account is active (not suspended)
//   4. Returns "Accept" (ResultCode: 0) or "Reject" (ResultCode: 1)
// 
// WHY THIS MATTERS:
//   If we return "Reject", Safaricom CANCELS the payment. 
//   Resident gets "Invalid account" on their phone. No lost money.
// ============================================================================
router.post('/validate', async (req, res) => {
  try {
    // req.body = the JSON data Safaricom sends us
    // BillRefNumber = the account number the resident typed (e.g., "348/4")
    // TransAmount = how much they want to pay (e.g., "400")
    // MSISDN = their phone number (e.g., "254712345678")
    const { BillRefNumber, TransAmount, MSISDN } = req.body;
    
    // Log the validation request so we can debug later
    console.log('M-PESA Validation Request:', req.body);
    
    // --- STEP 1: FIND THE HOUSE BY ACCOUNT NUMBER ---
    // We JOIN with residents to get the resident's name for the response
    // LEFT JOIN = include house even if no resident is linked (rare but possible)
    const houseResult = await db.query(
      `SELECT h.id, h.status, r.full_name, r.phone_number
       FROM houses h
       LEFT JOIN residents r ON r.house_id = h.id AND r.is_primary = true
       WHERE h.account_number = $1`,
      [BillRefNumber]
    );
    
    // --- STEP 2: CHECK IF ACCOUNT EXISTS ---
    // If no rows returned = account number doesn't exist in our database
    if (houseResult.rows.length === 0) {
      console.log(`Validation FAILED: Account ${BillRefNumber} not found`);
      // ResultCode: 1 = REJECT. Safaricom will show "Invalid account" to resident
      return res.json({ ResultCode: 1, ResultDesc: 'Invalid account number' });
    }
    
    // --- STEP 3: CHECK IF ACCOUNT IS ACTIVE ---
    // Status can be: 'active', 'inactive', 'suspended', 'vacant'
    // Only 'active' accounts can receive payments
    const house = houseResult.rows[0];
    
    if (house.status !== 'active') {
      console.log(`Validation FAILED: Account ${BillRefNumber} is ${house.status}`);
      return res.json({ ResultCode: 1, ResultDesc: 'Account is not active' });
    }
    
    // --- STEP 4: ACCEPT THE PAYMENT ---
    // ResultCode: 0 = ACCEPT. Safaricom proceeds to take the money
    console.log(`Validation SUCCESS: Account ${BillRefNumber} found for ${house.full_name}`);
    
    res.json({ 
      ResultCode: 0, 
      ResultDesc: 'Accepted',
      accountNumber: BillRefNumber,
      houseId: house.id
    });
    
  } catch (error) {
    // If our server crashes during validation, we STILL return "Accept"
    // Why? Because rejecting valid payments is worse than accepting invalid ones
    // We can always refund, but we can't undo a rejected payment
    console.error('Validation error:', error);
    res.json({ ResultCode: 1, ResultDesc: 'System error' });
  }
});

// ============================================================================
// ROUTE 2: M-PESA CONFIRMATION URL (POST /api/payments/confirm)
// ============================================================================
// WHEN DOES THIS FIRE?
//   After Safaricom successfully takes the money from the resident
//   This is the "receipt" — Safaricom tells us the payment is DONE
// 
// WHAT DOES IT DO?
//   1. Receives the M-PESA transaction details
//   2. Finds the house by account number
//   3. Finds the most recent unpaid bill for that house
//   4. Records the payment in our database
//   5. Updates the bill balance (trigger handles this automatically)
//   6. Sends SMS confirmation to the resident
// 
// WHY THIS MATTERS:
//   This is where the MAGIC happens. Zero manual work.
//   Faith pays → this endpoint fires → bill marked PAID → SMS sent → done.
// ============================================================================
router.post('/confirm', async (req, res) => {
  try {
    // Same fields as validation, PLUS:
    // TransID = the M-PESA receipt number (e.g., "ABC123XYZ")
    // This is what residents reference when they call support
    const { BillRefNumber, TransAmount, TransID, MSISDN, TransTime } = req.body;
    
    console.log('M-PESA Confirmation:', req.body);
    
    // --- STEP 1: FIND THE HOUSE ---
    const houseResult = await db.query(
      `SELECT h.id, r.full_name, r.phone_number
       FROM houses h
       LEFT JOIN residents r ON r.house_id = h.id AND r.is_primary = true
       WHERE h.account_number = $1`,
      [BillRefNumber]
    );
    
    // If account not found, we still return "Received" to Safaricom
    // This prevents Safaricom from retrying and flooding us
    if (houseResult.rows.length === 0) {
      return res.json({ ResultCode: 0, ResultDesc: 'Received' });
    }
    
    const house = houseResult.rows[0];
    
    // --- STEP 2: FIND THE MOST RECENT UNPAID BILL ---
    // We look for bills with status: 'pending', 'partial', or 'overdue'
    // ORDER BY created_at DESC = newest bill first
    // LIMIT 1 = only the most recent one
    const billResult = await db.query(
      `SELECT id, balance FROM bills 
       WHERE house_id = $1 AND status IN ('pending', 'partial', 'overdue')
       ORDER BY created_at DESC LIMIT 1`,
      [house.id]
    );
    
    // If no unpaid bill found, we still record the payment
    // This handles cases where resident overpays or pays early
    const billId = billResult.rows.length > 0 ? billResult.rows[0].id : null;
    
    // --- STEP 3: RECORD THE PAYMENT ---
    // We insert into the payments table with ALL the M-PESA details
    // This creates an audit trail — every KES is traceable
    const paymentId = uuidv4();
    await db.query(
      `INSERT INTO payments (id, bill_id, house_id, amount, mpesa_code, mpesa_phone, payment_method, status)
       VALUES ($1, $2, $3, $4, $5, $6, 'mpesa', 'completed')`,
      [paymentId, billId, house.id, parseFloat(TransAmount), TransID, MSISDN]
    );
    
    // --- STEP 4: CALCULATE NEW BALANCE ---
    // We sum all unpaid bills for this house
    // COALESCE = if no unpaid bills, return 0 instead of NULL
    const balanceResult = await db.query(
      `SELECT COALESCE(SUM(balance), 0) as total_balance FROM bills WHERE house_id = $1 AND status != 'paid'`,
      [house.id]
    );
    const newBalance = parseFloat(balanceResult.rows[0].total_balance) - parseFloat(TransAmount);
    
    // --- STEP 5: SEND SMS CONFIRMATION ---
    // We try to send SMS. If it fails (no API key, no internet), we log it but don't crash
    try {
      const smsService = require('../services/smsService');
      await smsService.sendPaymentConfirmation(
        house.phone_number,           // Resident's phone
        house.full_name || 'Resident', // Resident's name (fallback if null)
        BillRefNumber,                // Account number
        TransAmount,                  // Amount paid
        TransID,                      // M-PESA receipt code
        Math.max(0, newBalance)       // New balance (can't be negative)
      );
    } catch (smsError) {
      // Log the error but DON'T crash the payment processing
      // SMS failure is annoying, but payment recording is critical
      console.error('SMS failed:', smsError);
    }
    
    // --- STEP 6: ACKNOWLEDGE TO SAFARICOM ---
    // We MUST return ResultCode: 0, or Safaricom will keep retrying
    res.json({ ResultCode: 0, ResultDesc: 'Received' });
    
  } catch (error) {
    // Even if everything crashes, tell Safaricom we got it
    console.error('Confirmation error:', error);
    res.json({ ResultCode: 0, ResultDesc: 'Received' });
  }
});

// ============================================================================
// ROUTE 3: GET PAYMENTS FOR A HOUSE (GET /api/payments/house/:houseId)
// ============================================================================
// PROTECTED ROUTE — requires admin login (JWT token)
// Shows all payments made for a specific house
// Used by admin dashboard to see payment history
// ============================================================================
router.get('/house/:houseId', authenticate, async (req, res) => {
  try {
    // req.params.houseId = the UUID from the URL
    // e.g., /api/payments/house/550e8400-e29b-41d4-a716-446655440000
    const result = await db.query(
      `SELECT p.*, b.bill_period 
       FROM payments p
       LEFT JOIN bills b ON p.bill_id = b.id
       WHERE p.house_id = $1
       ORDER BY p.paid_at DESC`,
      [req.params.houseId]
    );
    res.json({ payments: result.rows });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// ============================================================================
// ROUTE 4: MANUAL PAYMENT ENTRY (POST /api/payments/manual)
// ============================================================================
// PROTECTED ROUTE — admin/clerk only
// For recording payments that did NOT come through M-PESA:
//   - Cash payments (resident pays driver or office directly)
//   - Bank transfers (resident sends money to company bank account)
//   - Other methods
// ============================================================================
router.post('/manual', authenticate, authorize('super_admin', 'admin', 'clerk'), async (req, res) => {
  try {
    // req.body contains all the payment details from the admin form
    const { house_id, amount, payment_method, payment_reference, bill_id } = req.body;
    const id = uuidv4();
    
    const result = await db.query(
      `INSERT INTO payments (id, bill_id, house_id, amount, payment_method, payment_reference, status)
       VALUES ($1, $2, $3, $4, $5, $6, 'completed') RETURNING *`,
      [id, bill_id, house_id, amount, payment_method, payment_reference]
    );
    
    res.status(201).json({ payment: result.rows[0] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// ============================================================================
// EXPORT THE ROUTER
// ============================================================================
// This makes all the routes above available to server.js
// server.js does: app.use('/api/payments', paymentRoutes)
// So '/validate' becomes '/api/payments/validate'
// ============================================================================
module.exports = router;

File: app/routers/payments.py
"""
SMART BIN LTD - PAYMENTS ROUTER (M-PESA + Manual)
"""
import uuid
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from app.database import get_db
from app.models import Payment, Bill, House, Phase, Estate, Resident
from app.schemas import PaymentCreate, MpesaValidationRequest, MpesaConfirmationRequest, PaymentOut
from app.dependencies import get_current_user, require_role
from app.services.mpesa_service import mpesa_service
from app.services.sms_service import sms_service
from app.config import get_settings

router = APIRouter(prefix="/api/payments", tags=["payments"])
settings = get_settings()

@router.post("/validate")
def mpesa_validate(req: MpesaValidationRequest, db: Session = Depends(get_db)):
    account_number = req.BillRefNumber.strip()
    amount = Decimal(req.TransAmount)

    house = db.query(House).filter(House.account_number == account_number).first()
    if not house:
        return {"ResultCode": 1, "ResultDesc": f"Account {account_number} not found"}
    if house.status != "active":
        return {"ResultCode": 1, "ResultDesc": f"Account {account_number} is inactive"}

    open_bills = db.query(Bill).filter(
        Bill.house_id == house.id,
        Bill.status.in_(["pending", "partial", "overdue"])
    ).all()

    if not open_bills:
        return {"ResultCode": 1, "ResultDesc": f"No open bills for account {account_number}"}

    total_due = sum(b.balance for b in open_bills)
    if amount > total_due * Decimal("1.5"):
        return {"ResultCode": 1, "ResultDesc": f"Amount {amount} exceeds total due {total_due}"}

    return {"ResultCode": 0, "ResultDesc": "Accepted"}

@router.post("/confirm")
def mpesa_confirm(req: MpesaConfirmationRequest, db: Session = Depends(get_db)):
    account_number = req.BillRefNumber.strip()
    amount = Decimal(req.TransAmount)
    mpesa_code = req.TransID
    phone = req.MSISDN

    house = db.query(House).filter(House.account_number == account_number).first()
    if not house:
        return {"ResultCode": 1, "ResultDesc": "Account not found"}

    bill = db.query(Bill).filter(
        Bill.house_id == house.id,
        Bill.status.in_(["pending", "partial", "overdue"])
    ).order_by(Bill.due_date.asc()).first()

    bill_id = bill.id if bill else None

    payment = Payment(
        id=str(uuid.uuid4()),
        bill_id=bill_id,
        house_id=house.id,
        amount=amount,
        mpesa_code=mpesa_code,
        mpesa_phone=phone,
        payment_method="mpesa",
        status="completed"
    )
    db.add(payment)
    db.commit()
    db.refresh(payment)

    if bill:
        db.refresh(bill)
        new_balance = float(bill.balance)
    else:
        new_balance = 0.0

    resident = db.query(Resident).filter(
        Resident.house_id == house.id,
        Resident.is_primary == True,
        Resident.is_active == True
    ).first()

    if resident:
        sms_service.send_payment_confirmation(
            phone=resident.phone_number,
            name=resident.full_name,
            account_number=house.account_number,
            amount=float(amount),
            mpesa_code=mpesa_code,
            balance=new_balance,
            house_id=house.id,
            db=db
        )

    return {"ResultCode": 0, "ResultDesc": "Success"}

@router.post("/register-urls")
def register_mpesa_urls(user=Depends(require_role("super_admin", "admin"))):
    validation_url = settings.mpesa_validation_url
    confirmation_url = settings.mpesa_confirmation_url

    if not validation_url or not confirmation_url:
        raise HTTPException(
            status_code=400,
            detail="MPESA_VALIDATION_URL and MPESA_CONFIRMATION_URL must be set in .env"
        )

    result = mpesa_service.register_urls(validation_url, confirmation_url)
    return {"message": "URLs registered", "result": result}

@router.post("/manual", response_model=PaymentOut)
def create_manual_payment(
    req: PaymentCreate,
    db: Session = Depends(get_db),
    user=Depends(require_role("super_admin", "admin", "clerk"))
):
    payment = Payment(
        id=str(uuid.uuid4()),
        bill_id=req.bill_id,
        house_id=req.house_id,
        amount=req.amount,
        payment_method=req.payment_method,
        payment_reference=req.payment_reference,
        status="completed"
    )
    db.add(payment)
    db.commit()
    db.refresh(payment)
    return payment

@router.get("/")
def list_payments(
    house_id: str = None,
    method: str = None,
    page: int = 1,
    limit: int = 50,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    offset = (page - 1) * limit

    q = db.query(
        Payment, Bill.bill_period.label("bill_period")
    ).outerjoin(Bill, Payment.bill_id == Bill.id)\
     .join(House, Payment.house_id == House.id)\
     .join(Phase, House.phase_id == Phase.id)\
     .join(Estate, Phase.estate_id == Estate.id)\
     .filter(Estate.company_id == user.company_id)

    if house_id:
        q = q.filter(Payment.house_id == house_id)
    if method:
        q = q.filter(Payment.payment_method == method)

    total = q.count()
    rows = q.order_by(Payment.paid_at.desc()).offset(offset).limit(limit).all()

    payments = []
    for p, period in rows:
        d = PaymentOut.model_validate(p)
        d.bill_period = period
        payments.append(d)

    from app.schemas import Pagination
    return {
        "payments": payments,
        "pagination": Pagination(
            page=page, limit=limit, total=total,
            total_pages=(total + limit - 1) // limit
        )
    }

@router.get("/house/{house_id}", response_model=list[PaymentOut])
def payments_by_house(house_id: str, db: Session = Depends(get_db), user=Depends(get_current_user)):
    return db.query(Payment).filter(Payment.house_id == house_id).order_by(Payment.paid_at.desc()).all()
